contrib/hyperscan/matching/input_matchers/source/matcher.cc

```diff
@@ -95,12 +75,82 @@ bool Matcher::match(absl::optional<absl::string_view> input) {
       },
       &matched);
   if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
-    ENVOY_LOG_MISC(error, "unable to scan, error code {}.", err);
+    IS_ENVOY_BUG(fmt::format("unable to scan, error code {}", err));
   }
 
   return matched;
 }
 
+std::string Matcher::replaceAll(absl::string_view value, absl::string_view substitution) const {
+  // Find matched bounds.
+  std::vector<Bound> bounds;
+  hs_scratch_t* scratch_ = tls_->get()->scratch_;
+  hs_error_t err = hs_scan(
+      som_database_, value.data(), value.size(), 0, scratch_,
+      [](unsigned int, unsigned long long from, unsigned long long to, unsigned int,
+         void* context) -> int {
+        std::vector<Bound>* bounds = static_cast<std::vector<Bound>*>(context);
+        bounds->push_back({from, to});
+
+        // Continue searching.
+        return 0;
+      },
+      &bounds);
+  if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
+    IS_ENVOY_BUG(fmt::format("unable to scan, error code {}", err));
+    return std::string(value);
+  }
+
+  // Sort bounds. Make sure the longest length bound in the front will appear first.
+  std::sort(bounds.begin(), bounds.end());
+
+  // Concatenate string and replace matched pair with substitution.
+  std::vector<absl::string_view> parts;
+  uint64_t pos = 0;
+  for (Bound& bound : bounds) {
+    if (bound.start_ < pos) {
+      continue;
+    }
+
+    parts.emplace_back(value.substr(pos, bound.start_ - pos));
+    parts.emplace_back(substitution);
+    pos = bound.end_;
+  }
+  parts.emplace_back(value.substr(pos));
+  return absl::StrJoin(parts, "");
+}
+
+bool Matcher::match(absl::optional<absl::string_view> input) {
+  if (!input) {
+    return false;
+  }
+
+  return static_cast<Envoy::Regex::CompiledMatcher*>(this)->match(*input);
+}
+
+void Matcher::compile(const std::vector<const char*>& expressions,
+                      const std::vector<unsigned int>& flags, const std::vector<unsigned int>& ids,
+                      hs_database_t** database) {
+  hs_compile_error_t* compile_err;
+  hs_error_t err =
+      hs_compile_multi(expressions.data(), flags.data(), ids.data(), expressions.size(),
+                       HS_MODE_BLOCK, nullptr, database, &compile_err);
+  if (err != HS_SUCCESS) {
+    std::string compile_err_message(compile_err->message);
+    int compile_err_expression = compile_err->expression;
+    hs_free_compile_error(compile_err);
+
+    if (compile_err_expression < 0) {
+      throw EnvoyException(fmt::format("unable to compile database: {}", compile_err_message));
```

Comment:
do we need a testcase for this case?

Reply:
It will only happen on lacking memories or something else on config time.

Reply:
RE "lacking memory" I would use IS_ENVOY_BUG which is a hint that the lack of coverage is OK

RE "something else" -- like what? Invalid input? We should test that.

Reply:
Fixed, I will change it to `IS_ENVOY_BUG`.

For example, invalid parameters (flags), which will not happen.